docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,205 @@
+# Unlinked Shaders And Relocatable Shader Elf
+
+# INTRODUCTION
+
+Unlinked and relocatable shaders were designed with the intent of allowing
+offline compilation of shaders that will be linked together during pipeline
+creation. Compiling shaders offline means that the they can be compiled without
+running the vulkan application. The shaders will also be compiled individually.
+This is advantagous when the same shader is used in multiple pipelines because
+it will only be compiled once. It can also reduce the compilation time while the
+application is running.
+
+# Unlinked shaders vs relocatable shaders
+
+The two concepts are similar, but there is a minor, and important, distinction.
+An unlinked shader is an LGC concept. LGC will generate an unlinked shader when
+[`m_unlinked`](../llpc/context/llpcPipelineContext.h) in the pipeline context is
+set. When possible, LGC will generate code that can be linked with other
+unlinked shaders and the pipeline state to create the pipeline. The link step is
+supposed to be relatively fast, and it is implemented in the `ElfLinkerImpl`
+class in [ElfLinker.cpp](../lgc/elfLinker/ElfLinker.cpp).
+
+A relocatable shader is an LLPC concept. The flag
+[`enableRelocatableShaderElf`](../include/vkgcDefs.h) in the pipeline options
+causes LLPC to use relocatable shaders when building the pipeline. When building
+with relocatable shaders, LLPC will create a new pipeline state object for each
+shader, stripping out as much as it can from the pipeline state. The new
+pipeline is passed to LGC, where it will be compiled as an unlinked shader. LLPC
+will then link all of those unlinked shaders together to create the pipeline.
+
+The key difference is that unlinked shaders are built with whatever information
+they have while relocatable shaders remove anything that can be handled at link
+time. For example, if LGC is asked to compile an unlinked shader and the
+pipeline state contains the user data nodes, then relocations will not be added
+for the descriptor loads. However, if LLPC is asked to compile a pipeline using
+relocatable shaders, the user data nodes will be present. LLPC will not pass
+that information to LGC when compiling the unlinked shaders, and relocations
+will be used.
+
+# Strategies for handling missing pipeline state
+
+## Relocations
+
+[Relocations in elf files](https://refspecs.linuxbase.org/elf/gabi4+/ch4.reloc.html)
+are the standard way to fill in missing information that is only available when
+linking object files. Relocations are often used to represent the address of a
+symbol to be accessed. Once the address of the symbol has been resolved, the
+linker can fill in the value. Unlinked shaders will create relocations for the
+absolute address of special symbols. These special symbols will not be given an
+actual address to be used, but their address will represent a value to be used.
+
```

Comment:
I am not super clear on the meaning of the last sentence. Did you mean something like "data symbols" where the symbol's value itself it used, as opposed to dereferenced in a load instruction?

Reply:
Yes that is one way of thinking about it.

Reply:
I've updated the sentence.  I hope it is clearer.